pkg/agent/controllers/agent/agent.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	go wait.Until(a.runServiceWorker, time.Second, stopCh)
+	wg.Wait()
+	klog.Info("Lighthouse agent workers started")
+	<-stopCh
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (a *Controller) enqueueService(obj interface{}) {
+	if key := a.getEnqueueKey(obj); key != "" {
+		klog.V(4).Infof("Enqueueing service %v for agent", key)
+		a.serviceWorkqueue.AddRateLimited(key)
+	}
+}
+
+func (a *Controller) getEnqueueKey(obj interface{}) string {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	if a.excludeNamespaces[ns] || excludeSvcList[name] {
+		return ""
+	}
+	return key
+}
+func (a *Controller) runServiceWorker() {
+	for a.processNextService() {
+
+	}
+}
+
+func (a *Controller) processNextService() bool {
+	obj, shutdown := a.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer a.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			a.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of object before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := a.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// already deleted Forget and return
+					a.serviceWorkqueue.Forget(obj)
+				} else {
+					// could be API error, so we want to requeue
+					logAndRequeue(key, a.serviceWorkqueue)
+				}
+				return fmt.Errorf("error retrieving service %s: %v", name, err)
+			}
+
+			return a.serviceUpdater(service, key)
+		})
+		if retryErr != nil {
+			return retryErr
+		}
+
+		a.serviceWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (a *Controller) handleUpdateService(old interface{}, newObj interface{}) {
+	// ServiceIp can't change without deleting service, so we only check for globalIp change
+	if key := a.getEnqueueKey(newObj); key != "" {
+		oldGlobalIp := getGlobalIpFromService(old.(*corev1.Service))
+		newGlobalIp := getGlobalIpFromService(newObj.(*corev1.Service))
+		if oldGlobalIp != newGlobalIp {
+			a.serviceWorkqueue.AddRateLimited(key)
+		}
+	}
+}
+
+func (a *Controller) serviceUpdater(service *corev1.Service, key string) error {
+	klog.V(2).Infof("Updating Service %s", key)
+	return nil
+}
+
+func (a *Controller) handleRemovedService(obj interface{}) {
+	var service *corev1.Service
+	var ok bool
+	var key string
+
+	if key = a.getEnqueueKey(obj); key == "" {
```

Comment:
You don't need to call this since you're not queueing.

Reply:
Should probably rename or split getEnqueueKey. If you notice, am also using getEnqueueKey to find out if we should process this service or not. If not, it returns emtpy key.

Reply:
ahh right. For the check, you just need the namespace which you can extract directly from the service `obj`.  This will also make it easer to use with the workqueue wrapper added by https://github.com/submariner-io/admiral/pull/45. We can migrate to use that later.

Reply:
Yeah, will refactor to separate the getKey from check.